massa-execution/src/vm.rs

```diff
@@ -1,31 +1,85 @@
 use std::sync::{Arc, Mutex};
 
-use massa_models::{Address, BlockId, Slot};
+use massa_models::address::AddressHashMap;
+use massa_models::{Address, Amount, BlockId, Slot};
 use tracing::debug;
 
 use crate::interface_impl::INTERFACE;
 use crate::sce_ledger::{SCELedger, SCELedgerChanges};
 use crate::types::{ExecutionContext, ExecutionStep, OperationSC, StepHistory};
-use crate::ExecutionConfig;
+use crate::{ExecutionError, ExecutionSettings};
 
 lazy_static::lazy_static! {
     pub(crate) static ref CONTEXT: Arc<Mutex::<ExecutionContext>> = {
-        let ledger = SCELedger::default(); // TODO Bootstrap
-        Arc::new(Mutex::new(ExecutionContext::new(ledger)))
+        let ledger = SCELedger::default();  // will be bootstrapped later
+        let ledger_at_slot = Slot::new(0, 0); // will be bootstrapped later
+        Arc::new(Mutex::new(ExecutionContext::new(ledger, ledger_at_slot)))
     };
 }
 
 pub struct VM {
-    _cfg: ExecutionConfig,
+    _cfg: ExecutionSettings,
     step_history: StepHistory,
 }
 
 impl VM {
-    pub fn new(_cfg: ExecutionConfig) -> VM {
-        VM {
-            _cfg,
-            step_history: Default::default(),
+    pub fn new(
+        cfg: ExecutionSettings,
+        thread_count: u8,
+        ledger_bootstrap: Option<(SCELedger, Slot)>,
+    ) -> Result<VM, ExecutionError> {
+        // bootstrap ledger
+        {
+            let context = CONTEXT.lock().unwrap();
+            let mut final_ledger_guard = context.ledger_step.final_ledger_slot.lock().unwrap();
+
+            if let Some((ledger_bootstrap, ledger_slot)) = ledger_bootstrap {
+                // bootstrap from snapshot
+                *final_ledger_guard = (ledger_bootstrap, ledger_slot);
+            } else {
+                // not bootstrapping: load initial SCE ledger from file
+                let ledger_slot = Slot::new(0, thread_count.saturating_sub(1)); // last genesis block
+                let ledgger_balances = serde_json::from_str::<AddressHashMap<Amount>>(
+                    &std::fs::read_to_string(&cfg.initial_sce_ledger_path).map_err(|err| {
+                        ExecutionError::FileError(format!(
+                            "error loading initial SCE ledger file {}: {}",
+                            cfg.initial_sce_ledger_path
+                                .to_str()
+                                .unwrap_or("(non-utf8 path)"),
+                            err
+                        ))
+                    })?,
+                )
+                .map_err(|err| {
+                    ExecutionError::FileError(format!(
+                        "error parsing initial SCE ledger file {}: {}",
+                        cfg.initial_sce_ledger_path
+                            .to_str()
+                            .unwrap_or("(non-utf8 path)"),
+                        err
+                    ))
```

Comment:
can you provide an example ?

Reply:
can you provide an example ? the 2 messages are not the same (one is "loading", the other "parsing")

Reply:
yep!

in error.rs for example you can have this:
```rust
macro_rules! bootstrap_error_while {
    ($st:expr, $cfg:ident) => {
        |err| {
            ExecutionError::FileError(format!(
                "error $st initial SCE ledger file {}: {}",
                $cfg.initial_sce_ledger_path
                    .to_str()
                    .unwrap_or("(non-utf8 path)"),
                err
            ))
        }
    };
}
pub(crate) use bootstrap_error_while;
```
and then your code can looks like this:
```rust
let ledgger_balances = serde_json::from_str::<AddressHashMap<Amount>>(
                &std::fs::read_to_string(&cfg.initial_sce_ledger_path)
                    .map_err(bootstrap_error_while!("loading", cfg))?,
            )
            .map_err(bootstrap_error_while!("parsing", cfg))?;
```

Reply:
I personally find this more complicated and much less clear/readable :(

But if you believe it's better, I trust your experience. Updated.
I just changed the name "bootstrap_error_while" into "bootstrap_file_error" because it's file-specific.